Add a "force" checkbox to the cancel pipeline modal dialog

jervi opened the pull request:
* Adds a `force=true` parameter to the backend request if checked
* Corresponds with the following PR’s in Gate and Orca:
  * https://github.com/spinnaker/orca/pull/1188
  * https://github.com/spinnaker/gate/pull/345
* The checkbox will only show if the execution engine is `v2`

duftler commented:
Can you please include a screen-grab on the PR?

jervi commented:
![Force checkbox](http://bildr.no/image/Z1BTZ3FU.jpeg)
![Checked](http://bildr.no/image/cFBwaVE0.jpeg)

anotherchrisberry commented:
The confirmation modal is meant to be general purpose. I'd much rather see this done by using a custom modal and controller from `execution.directive.js` than wedging it into the confirmation modal.

jervi commented:
You mean that we should not use the confirmation modal at all for the cancel functionality, but rather replace it with a custom modal?

anotherchrisberry commented:
Yeah, exactly. Let me know if you need help or have any questions when building it out.

jervi commented:
Any input on the UX btw?

danielpeach commented:
"cancel a pipeline using force" sounds a little off to me. What about "This option will force the pipeline to cancel."?

jervi commented:
Thank you for feedback!
@danielpeach:
![New text](http://bildr.no/image/TXJKUlkw.jpeg)

@anotherchrisberry I've reverted the changes to the confirmationModal, and implemented a new cancelModal. Any feedback is welcomed :)

icfantv commented on app/scripts/modules/core/cancelModal/cancelModal.controller.js:
```diff
@@ -0,0 +1,34 @@
+'use strict';
```
Can you please convert this to Typescript?

icfantv commented on app/scripts/modules/core/cancelModal/cancel.html:
```diff
@@ -0,0 +1,51 @@
+<div modal-page class="cancel-modal">
+  <form role="form" class="container-fluid no-padding" ng-submit="ctrl.confirm()">
+    <modal-close (dismiss)="$dismiss()"></modal-close>
```
We had to pull ng2 in a subsequent commit for performance reasons so this needs to be flipped back to ng1 syntax, i.e., just drop the `()` around the `dismiss` attribute.

icfantv commented on app/scripts/modules/core/cancelModal/cancel.html:
```diff
@@ -0,0 +1,51 @@
+        </div>
+      </div>
+      <task-reason command="params"></task-reason>
+
```
whitespace

icfantv commented on app/scripts/modules/core/cancelModal/cancel.html:
```diff
@@ -0,0 +1,51 @@
+          </div>
+        </div>
+      </div>
+
```
whitespace

icfantv commented on app/scripts/modules/core/cancelModal/cancelModal.controller.js:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    this.confirm = () => {
+      if (!this.formDisabled()) {
```
You will also need to check the submitting state here because it's possible for the UI to not be updated before the user [accidentally] clicks the button a second time - rendering your `ng-disabled` worthless.

jervi replied:
Thanks, but I'm not entirely sure what you mean by this. The `formDisabled` function is just `() => $scope.state.submitting;`. Do I have to check the submitting state another way as well?

icfantv commented on app/scripts/modules/core/cancelModal/cancelModal.service.ts:
```diff
@@ -0,0 +1,45 @@
+import {module} from 'angular';
+import {IModalService, IModalSettings} from 'angular-ui-bootstrap';
+
+export interface ICancelModalParams {
```
Are all these properties really optional?

icfantv commented on app/scripts/modules/core/cancelModal/cancelModal.service.ts:
```diff
@@ -0,0 +1,45 @@
+
+    return this.$uibModal.open(modalArgs).result;
+  }
+
```
whitespace

icfantv commented on app/scripts/modules/core/cancelModal/cancelModal.service.ts:
```diff
@@ -0,0 +1,45 @@
+
+    const modalArgs: IModalSettings = {
+      templateUrl: require('./cancel.html'),
+      controller: 'CancelModalCtrl as ctrl',
```
You can use the `controllerAs` property instead of making this inline.

jervi commented:
@icfantv Thank you for all the comments. Please have a look at the updated PR; especially the typescript conversion (I'm a TypeScript newbie :)

anotherchrisberry commented:
I think this is fine, just needs to be rebased and squashed into a single commit.

icfantv commented on app/scripts/modules/core/cancelModal/cancelModal.controller.ts:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  confirm(): void {
+    if (!this.formDisabled()) {
```
I meant that you should also have a check here for whether or not the form is already being submitted in your if statement.  Currently you are relying on the angular digest cycle to disable the button in your modal and you can't do this in a deterministic way.

jervi replied:
You mean like this: `if (!this.formDisabled() && !this.$scope.state.submitting) {`? I don't understand how that will make any difference, as the `formDisabled()` function just returns the value of `this.$scope.state.submitting`. You might argue that the whole `formDisabled` function could be replaced with submitting state though, both here and in the html...

icfantv replied:
GAH!  I'm sorry.  I totally missed the `()` on the check.  It's good.

jervi commented:
Thanks for the feedback @icfantv, @anotherchrisberry and @danielpeach ! Squashed and rebased.
